src/infinitempomatrix.jl

```diff
@@ -44,111 +44,191 @@ function InfiniteMPOMatrix(data::Vector{Matrix{ITensor}}, translator::Function)
+    end
+  end
+  return left_links, right_links
+end
+
+function convert_itensor_to_itensormatrix(tensor; kwargs...)
+  if order(tensor) == 3
+    return convert_itensor_3vector(tensor; kwargs...)
+  elseif order(tensor) == 4
+    return convert_itensor_33matrix(tensor; kwargs...)
+  else
+    error(
+      "Conversion of ITensor into matrix of ITensor not planned for this type of tensors"
     )
   end
-  IN = Hm[N + 1, N + 1] * onehot(T, dag(il0) => 1) * onehot(T, il => dim(il)) #I op in the bottom left of Hm
-  H, ir = directsum(H => ir, IN => il0; tags="Link,right") #1-D directsum to the put I in the bottom row, to the right of M[1]
+end
 
-  return H, il, ir
+"Build the projectors on the three parts of the itensor used to split a MPO into an InfiniteMPOMatrix"
+function build_three_projectors_from_index(is::Index; kwargs...)
+  old_dim = dim(is)
+  new_tags = get(kwargs, :tags, tags(is))
+  #Build the local projectors.
+  #We have to differentiate between the dense and the QN case
+  #Note that as far as I know, the MPO even dense is always guaranteed to have identities at both corners
+  #If it is not the case, my construction will not work
+  top = onehot(dag(is) => 1)
+  bottom = onehot(dag(is) => old_dim)
+  if length(is.space) == 1
+    new_ind = Index(is.space - 2; tags=new_tags)
+    mat = zeros(new_ind.space, is.space)
+    for x in 1:(new_ind.space)
+      mat[x, x + 1] = 1
+    end
+    middle = ITensor(copy(mat), new_ind, dag(is))
+  else
+    new_ind = Index(is.space[2:(end - 1)]; dir=dir(is), tags=new_tags)
+    middle = ITensors.BlockSparseTensor(
+      Float64,
+      undef,
+      Block{2}[Block(x, x + 1) for x in 1:length(new_ind.space)],
+      (new_ind, dag(is)),
+    )
+    for x in 1:length(new_ind.space)
+      dim_block = new_ind.space[x][2]
+      ITensors.blockview(middle, Block(x, x + 1)) .= diagm(0 => ones(dim_block))
+    end
+    middle = itensor(middle)
+  end
+  return top, middle, bottom
 end
 
-function find_all_links(Hms::InfiniteMPOMatrix)
-  is = inds(Hms[1][1, 1]) #site inds
-  ir = noncommonind(Hms[1][2, 1], is) #This op should have one link index pointing to the right.
-  #
-  #  Set up return array of 2-tuples
-  #
-  indexT = typeof(ir)
-  TupleT = NTuple{2,indexT}
-  inds_array = Vector{TupleT}[]
-  #
-  #  Make a dummy index
-  #
-  il0 = Index(ITensors.trivial_space(ir); dir=dir(dag(ir)), tags="Link,l=0")
-  #
-  #  Loop over sites and nonzero matrix elements which are linked into the next
-  #  and previous iMPOMatrix.  
-  #
-  for n in 1:nsites(Hms)
-    Hm = Hms[n]
-    inds_n = TupleT[]
-    lx, ly = size(Hm)
-    @assert lx == ly
-    for iy in (ly - 1):-1:1
-      ix = iy + 1
-      il = ix < lx ? commonind(Hm[ix, iy], Hms[n - 1][ix + 1, iy + 1]) : dag(il0)
-      ir = iy > 1 ? commonind(Hm[ix, iy], Hms[n + 1][ix - 1, iy - 1]) : il0
-      push!(inds_n, (il, ir))
+function convert_itensor_33matrix(tensor; leftdir=ITensors.In, kwargs...)
```

Comment:
Could you explain what these functions are doing?

Reply:
Added docstrings for these functions.
They are just converting the tensors of a MPO/iMPO into 3x3 (or 1x3 or 3x1) matrices of MPOs